Unstructured-IO/unstructured
# build(deps): avoid version conflicts

Addresses [#631](https://github.com/Unstructured-IO/unstructured/issues/631).

- Uses constraints to keep dependency versions more consistent.
- Moves all dependencies to `.in` files which are then ingested by `setup.py`.
- Adds script to check consistency of all extras.
- Adds consistency check to CI.

I should note that while it shouldn't be possible to cause a conflict between `base.txt` and any of the extras (because `base.txt` constrains all the extras) it is possible to get a conflict between two of the extras files. There are ways of trying to avoid that (like constraining each file by all the files that have already been processed before it in the order given in the `make pip-compile` target) but the ones I could think of seemed a little overwrought, and come with problems of their own. If a conflict arises, it should be flagged by CI or locally with `make check-deps`. When/if that happens, you can resolve the conflict by adding appropriate global constraints in `requirements/constraints.txt`.

Also note that if `fileA.in` is constrained by `fileB.txt`, then `fileB.in` should be compiled _before_ `fileA.in` in the `make pip-compile` target. Otherwise `fileA.in` will be compiled with the old version of `fileB.txt` which can cause conflicts or keep dependencies from being updated properly.

#### Testing:
- Verify `make check-deps` says deps are consistent.
- Change line 59 of `requirements/ingest-s3.txt` to:
```
urllib3==1.27.1
```
- Verify `make check-deps` produces an error.
